File: sakila-payment-service/src/main/java/com/example/sakila/module/rental/RentalService.java
package com.example.sakila.module.rental;

import com.example.sakila.event.bus.EventBus;
import com.example.sakila.exception.DataConflictException;
import com.example.sakila.exception.NotFoundException;
import com.example.sakila.module.rental.repository.RentalRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class RentalService {

  private final EventBus eventBus;

  private final RentalRepository rentalRepository;

  @Autowired
  public RentalService(@Qualifier("RentalEventBus") EventBus eventBus, RentalRepository rentalRepository) {
    this.eventBus = eventBus;
    this.rentalRepository = rentalRepository;
  }

  public Rental getRentalById(Long id) {
    if (id == null) return null;
    return rentalRepository.getRentalById(id);
  }

  public List<Rental> getRentalsByInventoryId(Long id) {
    if (id == null) return null;
    return rentalRepository.getRentalsByInventoryId(id);
  }

  public List<Rental> getRentalsByCustomerId(Long id) {
    if (id == null) return null;
    return rentalRepository.getRentalsByCustomerId(id);
  }

  public List<Rental> getRentalsByStaffId(Long id) {
    if (id == null) return null;
    return rentalRepository.getRentalsByStaffId(id);
  }

  public Rental createRental(Rental rental) {
    return rentalRepository.insertRental(rental);
  }

  public Rental updateRental(Long id, Rental source) {
    Rental target = getRentalById(id);
    if (target == null) throw new NotFoundException("Rental for ID " + id + " does not exist");

    target.setRentalDate(source.getRentalDate());
    target.setReturnDate(source.getReturnDate());
    target.setCustomer(source.getCustomer());
    target.setFilmId(source.getFilmId());
    target.setStoreId(source.getStoreId());
    target.setStaffId(source.getStaffId());

    return rentalRepository.updateRental(target);
  }

  public void deleteRental(Long id) {
    Rental rental = getRentalById(id);
    if (rental == null) throw new NotFoundException("Rental for ID " + id + " does not exist");

    try {
      rentalRepository.deleteRental(rental);
    } catch (DataIntegrityViolationException e) {
      throw new DataConflictException(e.getMessage(), e);
    }
  }
}

File: sakila-payment-service/src/main/java/com/example/sakila/module/rental/RentalService.kt
package com.example.sakila.module.rental

import com.example.sakila.event.bus.EventBus
import com.example.sakila.exception.DataConflictException
import com.example.sakila.exception.NotFoundException
import com.example.sakila.module.payment.PaymentService
import com.example.sakila.module.rental.event.RentalEventUtils
import com.example.sakila.module.rental.event.model.RentalCreatedEvent
import com.example.sakila.module.rental.event.model.RentalDeletedEvent
import com.example.sakila.module.rental.event.model.RentalUpdatedEvent
import com.example.sakila.module.rental.repository.RentalRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service

@Service
class RentalService @Autowired constructor(
    @Qualifier("RentalEventBus")
    private val eventBus: EventBus,

    private val rentalRepository: RentalRepository,

    private val paymentService: PaymentService
) {

  private val log = LoggerFactory.getLogger(RentalService::class.java)

  fun getRentalById(id: Long?): Rental? {
    if (id == null) return null
    return rentalRepository.getRentalById(id)
  }

  fun getRentalsByCustomerId(id: Long?): List<Rental>? {
    if (id == null) return null
    return rentalRepository.getRentalsByCustomerId(id)
  }

  fun getRentalsByStaffId(id: Long?): List<Rental>? {
    if (id == null) return null
    return rentalRepository.getRentalsByStaffId(id)
  }

  private fun generateCreatedEvent(rental: Rental) {
    val dto = RentalEventUtils.toDTO(rental)
    val event = RentalCreatedEvent(dto)
    eventBus.emit(event)
  }

  fun createRental(rental: Rental): Rental {
    log.info("Creating Rental")
    val result = rentalRepository.insertRental(rental)
    log.info("Created Rental id: ${result.id}")

    generateCreatedEvent(result)

    return result
  }

  private fun generateUpdatedEvent(rental: Rental) {
    val dto = RentalEventUtils.toDTO(rental)
    val event = RentalUpdatedEvent(dto)
    eventBus.emit(event)
  }

  fun updateRental(id: Long, source: Rental): Rental {
    val target = getRentalById(id)?: throw NotFoundException("Rental for ID $id does not exist")
    log.info("Updating Rental (ID: $id)")

    target.rentalDate = source.rentalDate
    target.returnDate = source.returnDate
    target.customer = source.customer
    target.filmId = source.filmId
    target.storeId = source.storeId
    target.staffId = source.staffId

    val result = rentalRepository.updateRental(target)

    generateUpdatedEvent(result)

    return result
  }

  private fun generateDeletedEvent(id: Long) {
    val event = RentalDeletedEvent(id)
    eventBus.emit(event)
  }

  private fun deletePayments(rentalId: Long)  {
    paymentService.getPaymentsByRentalId(rentalId)?.forEach { paymentService.deletePayment(it.id!!) }
  }

  fun deleteRental(id: Long) {
    val rental = rentalRepository.getRentalById(id)?: throw NotFoundException("Rental for ID $id does not exist")
    log.info("Deleting Rental (ID: $id)")

    try {
      deletePayments(id)
      rentalRepository.deleteRental(rental)
      generateDeletedEvent(id)
    } catch (e: DataIntegrityViolationException) {
      throw DataConflictException(e.message?: "Failed deleting rental with id: $id")
    }
  }
}
